MudTable: Retain selection when using ServerData

ralvarezing opened the pull request:
## Description

Prevents the loss of the current selection when a filter or page changes and ServerData is used to populate the table.
This was caused by #5455 because was not taken in consideration that when ServerData is used, the entire list is not available to the MudTable context (server provides only the current view for the table).

fixes #7535

## How Has This Been Tested?
Added unit test TableMultiSelectionTest9

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after, it its moving parts, please provide high quality gif, wemb or mp4 -->

https://github.com/MudBlazor/MudBlazor/assets/40799354/cbf983f4-fee2-406b-9389-02e1273b1636

## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

Commit:
Fixed selection cleared when filter or page change

Commit:
Added unit tests and change fix approach to an option configurable

Commit:
Merge remote-tracking branch 'MudOrigin/dev' into fix/table-multiselect-filtering

Commit:
fix test with new version

Commit:
Merge remote-tracking branch 'MudOrigin/dev' into fix/table-multiselect-filtering

Commit:
commit to trigger actions

Commit:
Merge remote-tracking branch 'MudOrigin/dev' into fix/table-multiselect-filtering

ScarletKuro commented:
@henon what do you think about this?

mikes-gh commented:
My expectation is that a filter is applied to the whole dataset and no of pages reduced if necessary.
It feels like adding a parameter to fix a bug. I get that it breaks behaviour, but we need to discuss whether adding a parameter is the right way to go.
Are there any mainstream use cases for filtering on the current page only?

ralvarezing commented:
> My expectation is that a filter is applied to the whole dataset and no of pages reduced if necessary. It feels like adding a parameter to fix a bug. I get that it breaks behaviour, but we need to discuss whether adding a parameter is the right way to go. Are there any mainstream use cases for filtering on the current page only?
> 
> If you had 90 records and your page size was 100 it should filter all the records. I don't think page size should change that.

To be clear, this proposed solution does not affect the filtering/paging; what it does is to allow the programmer to preserve the selection when page or any filter changes. 
Also, the idea behind this is to prevent that the current selection (when the table is multiselect) is lost when filters or page changes.

To give a little more context: the problem with the current implementation is that when a MudTr (row in the component) is disposed, the row is removed from the Table and if selected also removed from the selection (solving #5455).
But this causes that every time the dataset is changed (ex: using ServerData and filtering at server side) the selection is lost (causing #7535).

I am also a user of this library, and this is a problem I usually face; specially when I need to allow the user to select items from more than one page, or let the user apply a filter to find more options.

henon commented:
I like it, very useful. I think that in other Frameworks such as in Vuetify this is even the default behavior.

henon commented:
@ralvarezing Could modify one of the MudTable examples in the docs so that is shows this new option and add a sentence to the description explaining it? Maybe here?

![image](https://github.com/MudBlazor/MudBlazor/assets/44090/2489d05e-dde9-43c2-8274-3901de6e1099)

Or maybe you find a more suitable example. The number of examples is already very high for mudtable so I'd not add a complete new one.

mikes-gh commented:
> To be clear, this proposed solution does not affect the filtering/paging; what it does is to allow the programmer to preserve the selection when page or any filter changes. Also, the idea behind this is to prevent that the current selection (when the table is multiselect) is lost when filters or page changes.

My apologies I read it wrong.

> I like it, very useful. I think that in other Frameworks such as in Vuetify this is even the default behavior.

Should we have it as default?
Does Vuetify have a parameter for this behaviour?
Do we need to make it configurable?

henon commented:
@mikes-gh I would love to make this the default behavior. And AFAIK Vuetify has no param for it. 

Does anyone have concerns that this might be breaking for some users who maybe rely on the current behavior? There is always the possibility to just see this as a bug that got fixed. We could keep the parameter for those who might rely on the old behavior so that they can switch it off.

Yomodo commented:
We would be very happy to see this implemented as this issue has been a UI bug in our backlog for a long time.
I could not think of a scenario to keep the old (bug) behavior alive.

ralvarezing commented:
> @ralvarezing Could modify one of the MudTable examples in the docs so that it shows this new option and add a sentence to the description explaining it? Maybe here?
> 
> ![image](https://private-user-images.githubusercontent.com/44090/306257233-2489d05e-dde9-43c2-8274-3901de6e1099.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0MzM0MjYsIm5iZiI6MTcwODQzMzEyNiwicGF0aCI6Ii80NDA5MC8zMDYyNTcyMzMtMjQ4OWQwNWUtZGRlOS00M2MyLTgyNzQtMzkwMWRlNmUxMDk5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIwVDEyNDUyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMzN2MyMzYxYWRmNmYyZThhYWEzY2M3Nzc4YWJhNzU4ZmU1YTg5NGM0MzU1ZjFiYmE4YWUyMmY2YWEyZjQ5MzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.V5PcMt1kF15aQBIBzrZg_Y_UGpRbSrKCjQxJfhf6AdU)
> 
> Or maybe you find a more suitable example. The number of examples is already very high for mudtable so I'd not add a complete new one.

Perfect! this is a great place to put an example of this. I'll see to add an example later.

mikes-gh commented:
My opinion is that if we add a parameter and make the default breaking, we are already making a breaking change.
It's a bug fix IMO and doesn't need API unless someone can come up with a scenario that they are using the bugs behaviour and can't fix their code in another way.

henon commented:
@ralvarezing Wouldn't you say that elements should be removed from the selection when deleting rows even with the new behavior? I guess the bug fix that tried to address this did just take an easy route and cleared the selection on any update. Maybe we could keep elements in the selection as long as they are still in the entire list of elements? If we can manage this I would say we don't even need the parameter. But if someone has good arguments for it it should definitiely be renamed like you suggested.

henon commented:
> Due that there was a reported and fixed bug to update the selection when removing an item [...]

@ralvarezing if you could link this issue and PR here and tag the reporter/author we could involve them in the discussion to gain more insights.

ralvarezing commented:
> @ralvarezing Wouldn't you say that elements should be removed from the selection when deleting rows even with the new behavior? I guess the bug fix that tried to address this did just take an easy route and cleared the selection on any update. Maybe we could keep elements in the selection as long as they are still in the entire list of elements? If we can manage this I would say we don't even need the parameter. But (like Mike said) if someone has good arguments for it, it should definitiely be renamed like you suggested.

I agree with what you said, 
but the problem shows when in the table context you don't have the entire list.
For example, when using ServerData to populate the table (for server-side paging/filtering) or having a filter outside the MudTable context.

In those cases, I don't see a way to know if the item was removed from the list or if the user changed page.
The programmer is the one that could know if is required to remove items from the selection or not.


@Mr-Technician could you give more information about this bug you reported #5455 and this pr?

Mr-Technician commented:
@ralvarezing I think the description of #5455 covers it pretty well. My PR ensured that when an item was removed from the table, it was also removed from the selection.

It does not make sense for a selection to contain items that have been removed from the table, however I could understand the selection containing items that still exist in the table but have been filtered out. At a minimum, the selection should retain items included in the filter.

henon commented:
@mr.technician I think when you use ServerData in MudTable and switch between pages the table will always only see the items of the current page because it requests every page from the server. It is this scenario that collides with your fix if one wants to switch between pages and keep the selection in tact.

henon commented:
As far as I understand, the problem which @Mr-Technician fixed in #5455 was that the check-all checkbox was behaving weird because of the selection containing elements that are outside of the current ServerData view. So I think the check-all checkbox state must be calculated based on the current ServerData view and all elements of the selection that are not in it must not affect its state, effectively using the intersection set of the current view and the selected items for determination of the check-all checkbox state instead of the entire selection.
So for example if you select all elements on page 1 it becomes checked, but when you go to page 2 it becomes unchecked because the selected element is not part of the current view. Does this make sense? 

I think if we can fix the check-all checkbox for ServerData in this way we won't need a parameter and can just switch to the new behavior.

ralvarezing commented:
That could be a good solution.
This way the behavior of the check-all is preserved and don't lose the current selection when changing pages/filters.

The only catch would be that being in page 2 with items selected in page 1, and don't having the check-all in the "partial selected state", from the user perspective could mean that there are no selected items and potentially lead to unwanted results.
This is a trade-off that could be ok depending on the system being developed.

Another posible option:
Prevent the selection from being cleared when ServerData is being used, where is alleged to be already filtered and paginated ("Set ServerData to load data from the backend that is filtered, sorted and paginated" from the documentation). 
In this case, the selection will be cleared only when using 'Items' and not 'ServerData'.

Unless there is a scenario where you don't want the selection to be preserved using ServerData, which I don't currently see.

The parameter won't be required in eigher option.

henon commented:
I may not understand exactly what you propose but I see no difference between option 1 and option 2 with respect to unchecking the check-all box. 

Of course with Items the check-all would still operate on all items. With ServerData in option 1 the check-all operates only on the currently shown items (paginated and/or filtered). So unchecking the check-all would not clear the entire selection but only the items in the selection that are on the currently viewed page.

Sure it is inconvenient that the user won't see that there are items selected on different pages. But I guess this is a visualization problem that doesn't have to be solved in this PR. We can add a warning to the ServerData documentation that the check-all checkbox operates only on the current view for now.

ralvarezing commented:
Actually, reading again your comment before, I see that they are the same option; the difference is that I propose to keep clearing the selection when not using ServerData and leave the calculation of the check-all unchanged (then the visual bug is not introduced).
The fix of #5455 only makes sense when using Items (not ServerData) because you don't have a 'remove item' when the data will be fetched from the server; so, the bugfix of #5455 should not be affected by this.

henon commented:
Yes, sorry I wasn't clear about the difference between using Items and ServerData we have to make. Great, I think we have a winner.

ralvarezing commented:
Great! then I will be updating this PR and develop this as a fix.

Commit:
Removed RettainSelection parameter & fix test

1 The Retain parameter was removed
2 If using ServerData, the Selection will not be cleared if the data is refreshed, to prevent lost of selection
3 The parameter is fixed to correlate new implementation

Commit:
Merge remote-tracking branch 'MudOrigin/dev' into fix/table-multiselect-filtering

ralvarezing commented:
@henon ready for review!

my mistake not switching back to draft before :D

henon reviewed (approved):
Wow, the solution is surprisingly simple, which is great. Thank you very much!
